Fix nodes has more than one output users after replace_set_grad_with_hop pass

ydwu4 opened the pull request:
Summary: Previously, when we inline the subgraphs that doesn't have a different require_grad environment, we didn't clean up the nodes's users in subgraph and direcly used them to  to replace the output of  the call_modules. This records dead depencies in node.users. This PR fixes this.

Test Plan:
Added a new test.

Also see the torchrec tests:
Step 1:
buck run mode/dev-nosan //aimp/experimental/pt2:pt2_export -- --model-entity-id 934687114 --output /tmp/934687114.zip --use-torchrec-eager-mp --use-manifold

Step 2:
buck run mode/opt -c python.package_style=inplace -c fbcode.enable_gpu_sections=true aimp/cli:cli --  --platform=aps --template=disagg_gpu_aps_pt2 --pt2 --model-entity-id=934687114 non-request-only-tagging torchrec-shard-and-quantize gpu-disagg-split assign-device materialize-weights script-and-save

Differential Revision: D59132214

angelayi commented on torch/_export/passes/replace_set_grad_with_hop_pass.py:
```diff
@@ -140,51 +140,48 @@ def _sequential_split_and_maybe_inline_subgraphs(
-            )
-        new_gm.recompile()
-        return new_gm, new_signature
+    need_replacing = any(_is_set_grad_enabled_node(node) for node in gm.graph.nodes)
```
are there any logic changes in this block? it seems the same to me?

ydwu4 replied:
Yeah, same logic, just some code motion to make it looks clearer.

ydwu4 committed:
Fix nodes has more than one output users after replace_set_grad_with_hop pass (#129716)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/129716

Previously, when we inline the subgraphs that doesn't have a different require_grad environment, we didn't clean up the nodes's users in subgraph and direcly used them to  to replace the output of  the call_modules. This records dead depencies in node.users. This PR fixes this.

Test Plan:
Added a new test.

Also see the torchrec tests:
Step 1:
buck run mode/dev-nosan //aimp/experimental/pt2:pt2_export -- --model-entity-id 934687114 --output /tmp/934687114.zip --use-torchrec-eager-mp --use-manifold

Step 2:
buck run mode/opt -c python.package_style=inplace -c fbcode.enable_gpu_sections=true aimp/cli:cli --  --platform=aps --template=disagg_gpu_aps_pt2 --pt2 --model-entity-id=934687114 non-request-only-tagging torchrec-shard-and-quantize gpu-disagg-split assign-device materialize-weights script-and-save

Reviewed By: angelayi

Differential Revision: D59132214
